### PR TITLE
TRAP: Add functionality to dump a file from userfs partition on crash

### PR DESCRIPTION
**Pull request by amogh-samsung**
- This change allows dumping of a particular file by name/path from a crashed TizenRT device

Signed-off-by: Amogh Hassija <a.hassija@samsung.com>

USAGE: - After TizenRT device crashes

> 
> ************************************************************************************************
>                 Disconnect this serial terminal and run Dump Tool
> ************************************************************************************************
> 
> 

> /TizenRT/tools/trap$ sudo ./trap.sh 
> Please enter serial port adapter:
> For example: /dev/ttyUSB1 or /dev/ttyACM0
> Enter:
> /dev/ttyUSB1
> Target device locked and ready to DUMP!!!
> Choose from the following options:-
> 1. RAM Dump
> 2. Userfs Dump
> 3. Both RAM and Userfs dumps
> 4. External Userfs Partition Dump
> 5. Dump specific file from target device
> r. Reboot TARGET Device
> x. Exit TRAP tool
> You may chose mutiple dump options together by entering the option numbers consectuively
> For Example - '14' to dump both RAM and External USERFS partition contents OR '35' to print RAM, USERFS and FILE contents
> 5
> This utility will allow you to enter a file name that you want to be dumped from the device
> It is the responsibility of the user to ensure that the correct file name/path is provided
> do_handshake: Target Handshake successful
> Enter the absolute path of the file that you want to be dumped from the Filesystem: -/mnt/f1.txt
> Attempting to Dump File /mnt/f1.txt from Target Device
> 
> Receiving file /mnt/f1.txt of size 5522 from Target Device
> 
> [>]
> 
> Dumped file /mnt/f1.txt from filesystem successfully
> Choose from the following options:-
> 1. RAM Dump
> 2. Userfs Dump
> 3. Both RAM and Userfs dumps
> 4. External Userfs Partition Dump
> 5. Dump specific file from target device
> r. Reboot TARGET Device
> x. Exit TRAP tool
> You may chose mutiple dump options together by entering the option numbers consectuively
> For Example - '14' to dump both RAM and External USERFS partition contents OR '35' to print RAM, USERFS and FILE contents
> r
> NOTE: - CONFIG_BOARD_ASSERT_AUTORESET needs to be enabled to reboot TARGET Device after a crash
> Handsake for rebooting device may fail since device reboots before sending acknowledgement
> Wrong Target Handshake, ack = H
> Target Handshake Failed
> Dump tool exits after successful operation

### CODE REVIEW

File: tools/trap/trap.c [hunk 1]
```diff
@@ -288,6 +291,67 @@ static int extfsdump_recv(int dev_fd)
...
 }
 
+int dump_file_from_fs(int dev_fd, char *filename)
+{
+	int ret;
+	int fd;
+	int i, last_subdir;
+	char buf;
+	FILE *fp;
+	u_int64_t file_size;
```
Comment (thapav) [inline tools/trap/trap.c:301]: uint64_t*

File: os/board/common/crashdump.c [hunk 2]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+
+	ret = stat(filename, &st);
+	if (ret != OK) {
+		/* We are not able to stat the file for some reason */
+		return -1;
+	}
+	size = st.st_size;
+
+	/* This Ack signal confirms the presence of the requested file */
+	//up_lowputc('A');
```
Comment (thapav) [inline os/board/common/crashdump.c:240]: Is this needed ? If not please remove

File: os/board/common/crashdump.c [hunk 3]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+	 * This also acts as a confirmation of the presence of the file in itself.
+	 */
+	for (i = 0; i < 8; i++) {
+		char_filesize = (size >> ((8 - 1 - i) * 8));
+		up_lowputc(char_filesize);
+	}
+
+	fd = open(filename, O_RDOK);
+	if (fd < 0) {
+		//Unable to open file for transmitting data
```
Comment (thapav) [inline os/board/common/crashdump.c:253]: Please maintain consistent comment style
/* */

File: os/board/common/crashdump.c [hunk 4]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+		return -EIO;
+	}
+
+	while (size) {
+		ret = read(fd, &buf, 1);
+		if (ret != 1) {
+			//Unable to read from file
```
Comment (thapav) [inline os/board/common/crashdump.c:260]: Same as above

File: os/board/common/crashdump.c [hunk 5]
```diff
@@ -450,6 +511,12 @@ static int dump_via_uart(void)
 			lldbg("Unable to reset device\n");
 #endif
 			break;
+		/* Check handshake string for USERFS FILE DUMP signal */
+		} else if (strncmp(host_buf, USERFS_FDUMP_HANDSHAKE_STRING, strlen(USERFS_FDUMP_HANDSHAKE_STRING)) == 0) {
```
Comment (thapav) [inline os/board/common/crashdump.c:515]: Kinldy use strnlen()

**Comment by sunghan-chang** [comment]
@amogh-samsung Could you add some example how to use in the commit description?

File: tools/trap/trap.c [hunk 6]
```diff
@@ -24,6 +24,9 @@
  ****************************************************************************/
 
 #include "trap.h"
+#include "../../os/include/tinyara/config.h"
```
Comment (sunghan-chang) [inline tools/trap/trap.c:27]: Could you let me know why includes this?

Comment (amogh-samsung) [reply tools/trap/trap.c:27]: It is used for `CONFIG_SMARTFS_MAXNAMLEN`
The maximum possible size of the file path string

Comment (sunghan-chang) [reply tools/trap/trap.c:27]: 1. But there is a definition for that at line 29.
[CONFIG_SMARTFS_MAXNAMLEN](https://github.com/Samsung/TizenRT/pull/5738/files#diff-d50c5afb6cd42629b54e5c4ead10ed2e4057df258a2e1f16d7a8f09a326679e9R29)
2. If user does not have built source, then this will be failed. Do we need to build TizenRT to use TRAP?

Comment (amogh-samsung) [reply tools/trap/trap.c:27]: Sorry I need to remove the one in line 29
And I agree with point 2 given by you

Would you suggest I continue with a manually defined dummy config for the purpose of this tool?

Comment (sunghan-chang) [reply tools/trap/trap.c:27]: Thapa also reads some config value on the TRAP. Could you check it with her about how to get a config value? Let's use the same way.

Comment (amogh-samsung) [reply tools/trap/trap.c:27]: I checked tools/trap/ramdumpParser.py
This also uses the TizenRT .config file (Line 1075)

How about adding a comment in HowToUseTrap.md file about the need for building source code for making the config file available?

Comment (sunghan-chang) [reply tools/trap/trap.c:27]: I think 
1. trying to read .config file from default location
2. If no .config there, feedback an error mesage about no .config and wait an input of configuration file 

Many developer uses the tool without source code so that I think getting necessary file separately is better.

File: os/board/common/crashdump.c (hunk 5 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c:515]: All handshake checks are currently using `strlen`, so sticking with the same.

File: tools/trap/trap.c (hunk 6 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:27]: Pending, will take care of this

File: tools/trap/trap.c (hunk 1 above)
Comment (amogh-samsung) [reply tools/trap/trap.c]: Done

File: os/board/common/crashdump.c (hunk 2 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c]: Done

File: os/board/common/crashdump.c (hunk 3 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c]: Done

File: os/board/common/crashdump.c (hunk 4 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c]: Done

**Comment by sunghan-chang** [comment]
> RAM Dump
Userfs Dump
Both RAM and Userfs dumps
External Userfs Partition Dump
Exit TRAP Tool
Reboot TARGET Device
Dump specific file from Target Device

The option after exit looks strange. How about adding that before exit and reboot?
To do this, we had better modify the numbers for exit and reboot not to get affected from new adding like
```
r. Reboot TARGET device
x. Exit TRAP
```

**Comment by amogh-samsung** [comment]
> > 1. RAM Dump
> > 2. Userfs Dump
> > 3. Both RAM and Userfs dumps
> > 4. External Userfs Partition Dump
> > 5. Exit TRAP Tool
> > 6. Reboot TARGET Device
> > 7. Dump specific file from Target Device
> 
> The option after exit looks strange. How about adding that before exit and reboot? To do this, we had better modify the numbers for exit and reboot not to get affected from new adding like
> 
> ```
> r. Reboot TARGET device
> x. Exit TRAP
> ```

I agree, this makes sense
But perhaps as a separate commit maybe in the same or different PR?
Would that be alright?

**Comment by sunghan-chang** [comment]
> I agree, this makes sense
But perhaps as a separate commit maybe in the same or different PR?
Would that be alright?

Both. Anything is ok for me.

File: tools/trap/trap.c (hunk 6 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:27]: I pushed a slightly different version.
My estimate is that if the config.h file is not present because the user did not build the code, it is unlikely that they would have the config.h or config file present separately.
Even if they do, I feel it is better to ask the user for the config value directly instead of the file path.
The value is this config is always (so far in all defconfigs) = 32

Also, dynamic inclusion of the provided file path as a header file in C would not be easy.

File: tools/trap/trap.c [hunk 7]
```diff
@@ -288,6 +289,67 @@ static int extfsdump_recv(int dev_fd)
...
+	}
+
+	file_size = (((u_int64_t)char_filesize[0] << 56) & 0xFF00000000000000U) | (((u_int64_t)char_filesize[1] << 48) & 0x00FF000000000000U) | (((u_int64_t)char_filesize[2] << 40) & 0x0000FF0000000000U) | (((u_int64_t)char_filesize[3] << 32) & 0x000000FF00000000U) | ((char_filesize[4] << 24) & 0x00000000FF000000U) | ((char_filesize[5] << 16) & 0x0000000000FF0000U) | ((char_filesize[6] <<  8) & 0x000000000000FF00U) | (char_filesize[7] & 0x00000000000000FFU);
+
+	printf("\nReceiving file %s of size %lu from Target Device\n\n", filename, file_size);
+
+	i = 0;
+	last_subdir = -1;
+	while (filename[i] != '\0') {
+		/* We dont want the file to saved in the root folder, we want it in the current folder */
```
Comment (abhishek-samsung) [inline tools/trap/trap.c:327]:
```suggestion
		/* We dont want the file to be saved in the root folder, we want it in the current folder */
```

File: tools/trap/trap.c [hunk 8]
```diff
@@ -463,6 +528,40 @@ int main(int argc, char *argv[])
...
+			printf("This utility will allow you to enter a file name that you want to be dumped from the device\n");
+			printf("It is the responsibility of the user to ensure that the correct file name/path is provided\n");
+
+			strncpy(handshake_string, USERFS_FDUMP_HANDSHAKE_STRING, HANDSHAKE_STR_LEN_MAX);
+			if (do_handshake(dev_fd, handshake_string) != 0) {
+				printf("Target handshake failed\n");
+				goto handshake_err;
+			}
+
+			printf("Enter the path of the file that you want to be dumped from the Filesystem: -");
```
Comment (abhishek-samsung) [inline tools/trap/trap.c:551]: Is it **absolute path** or path relative to **/mnt**  ? I think its better to mention here to be clear.

Comment (amogh-samsung) [reply tools/trap/trap.c:551]: Let's mention absolute

File: tools/trap/trap.c (hunk 7 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:327]: Let me do this, thanks

File: os/board/common/crashdump.c [hunk 9]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+	 */
+	for (i = 0; i < 8; i++) {
+		char_filesize = (size >> ((8 - 1 - i) * 8));
+		up_lowputc(char_filesize);
+	}
+
+	fd = open(filename, O_RDOK);
+	if (fd < 0) {
+		//Unable to open file for transmitting data
+		return -EIO;
```
Comment (Taejun-Kwon) [inline os/board/common/crashdump.c:254]: Line 235 it returns -1, but here it returns -EIO.
Actually -1 is EPERM, so you should make clear this.
1. ERROR or OK
2. errno or OK

File: os/board/common/crashdump.c [hunk 10]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+	}
+
+	while (size) {
```
Comment (Taejun-Kwon) [inline os/board/common/crashdump.c:257]: You already know less I/O always improve performance.
```
char *buf = malloc(size...);
ret = read(fd, buf, size);
....
while (*size) {
   up_lowputc(*size++);
}
```

File: tools/trap/trap.c [hunk 11]
```diff
@@ -288,6 +289,67 @@ static int extfsdump_recv(int dev_fd)
...
+	file_size = 0;
+	for (int i = 0; i < 8; i++) {
+		ret = read(dev_fd, &char_filesize[i], sizeof(char_filesize[i]));
+		if (ret != sizeof(char_filesize[i])) {
+			printf("Unable to get file size from Target Device\n");
+			return -1;
+		}
+	}
+
+	file_size = (((u_int64_t)char_filesize[0] << 56) & 0xFF00000000000000U) | (((u_int64_t)char_filesize[1] << 48) & 0x00FF000000000000U) | (((u_int64_t)char_filesize[2] << 40) & 0x0000FF0000000000U) | (((u_int64_t)char_filesize[3] << 32) & 0x000000FF00000000U) | ((char_filesize[4] << 24) & 0x00000000FF000000U) | ((char_filesize[5] << 16) & 0x0000000000FF0000U) | ((char_filesize[6] <<  8) & 0x000000000000FF00U) | (char_filesize[7] & 0x00000000000000FFU);
```
Comment (Taejun-Kwon) [inline tools/trap/trap.c:320]: Instead of this tricky code, you can use uint64_t buf when you write & read.
write file size : write(fd, &buf, sizeof(uint64_t));
get file size : read(fd, &buf, sizeof(uint64_t));
anyway, is file_size written in header of file?

File: tools/trap/trap.c [hunk 12]
```diff
@@ -288,6 +289,67 @@ static int extfsdump_recv(int dev_fd)
...
+	filename = filename + last_subdir + 1;
+	snprintf(local_filename, CONFIG_SMARTFS_MAXNAMLEN + 9, "%s%s", "DumpFile_", filename);
+
+	fp = fopen(local_filename, "w");
+	if (fp == NULL) {
+		printf("Unable to open file %s to save logs chain dump\n", filename);
+		return -1;
+	}
+
+	ret = get_dump(dev_fd, fp, file_size);
```
Comment (Taejun-Kwon) [inline tools/trap/trap.c:343]:
```
get_dump(FILE *fp)
{
    fopen();
    ...do_something
    fclose();
}
```
get_dump seems that get data from somewhere, so it should handle fp alone

File: tools/trap/trap.c [hunk 13]
```diff
@@ -463,6 +528,40 @@ int main(int argc, char *argv[])
 			printf("\nExternal Userfs partition dump received successfully\n");
 			fflush(stdout);
 		}
+
+		if (DUMP_FLAGS & FILE_DUMP_FLAG) {
+#ifdef CONFIG_SMARTFS_MAXNAMLEN
```
Comment (Taejun-Kwon) [inline tools/trap/trap.c:533]: If I were you,

at the top of code,
```
#ifndef CONFIG_SMARTFS_MAXNAMLEN
#define CONFIG_SMARTFS_MAXNAMELEN 32
#endif
```
and if user input greater than this config, print out error message.

File: os/board/common/crashdump.c (hunk 10 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c:257]: I see crashdump.c does not use malloc at any place.
Also, are we really concerned about I/O time/performance when device is crashed? Our primary aim should be that we are able to extract the file/data correctly from the crashed device for analysis.

File: tools/trap/trap.c [hunk 14]
```diff
@@ -463,6 +528,40 @@ int main(int argc, char *argv[])
...
+			int smartfs_maxnamelength = CONFIG_SMARTFS_MAXNAMLEN;
+#else
+			int smartfs_maxnamelength;
+			printf("No Config found for maximum possible file name length, please enter manually (usually = 32)\n");
```
Comment (swasthik22) [inline tools/trap/trap.c:537]:
```suggestion
			printf("No Config found for maximum possible length of file name, please enter manually (usually = 32)\n");
```

File: tools/trap/trap.c [hunk 15]
```diff
@@ -288,6 +289,67 @@ static int extfsdump_recv(int dev_fd)
...
+		printf("Unable to send Dump File name to Target Device\n");
+		return -1;
+	}
+
+	file_size = 0;
+	for (int i = 0; i < 8; i++) {
+		ret = read(dev_fd, &char_filesize[i], sizeof(char_filesize[i]));
+		if (ret != sizeof(char_filesize[i])) {
+			printf("Unable to get file size from Target Device\n");
+			return -1;
```
Comment (swasthik22) [inline tools/trap/trap.c:316]: please check for consistency of return statements in all places. Some places there is flag based return, and in some there is numeric returns.

Comment (amogh-samsung) [reply tools/trap/trap.c:316]: This is already under discussion with regards to line 254 in one of the above conversations.
Please check that

File: tools/trap/trap.c (hunk 14 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:537]: Will do

File: tools/trap/trap.c (hunk 12 above)
Comment (Taejun-Kwon) [reply tools/trap/trap.c:343]: It is better that handle open & close in one function, it is always better option. you may know it enough, why don't you modify it?

File: tools/trap/trap.c (hunk 11 above)
Comment (Taejun-Kwon) [reply tools/trap/trap.c:320]: Because you write file name and readback, then calculate file size.
That's why asking you is there information regarding file size at the start of file.
I can't understand why this code as added because we can use stat to get size of file easily.

File: tools/trap/trap.c (hunk 13 above)
Comment (Taejun-Kwon) [reply tools/trap/trap.c:533]: Yes, maybe user who using dump may know well about our system & platform I agree with you. So,
1. Then it should ask size of name length to user all the time if you consider such a user as you mentioned above.
2. We can mount another file system, so here you should not refer to config in specific file system, smartfs.
3. But In my opinion, it should not check length of user input here, and fs will return ERROR & ENAMETOOLONG or EEXIST as a errno, then print out information regarding error when it fail to created file. Because they know well about our system as we expected.

File: tools/trap/trap.c (hunk 12 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:343]: Yes, it can be done.
I will check and create a different commit for that if confirmed

File: os/board/common/crashdump.c (hunk 10 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c:257]: @kishore-sn what is your opinion on this please?

File: tools/trap/trap.c (hunk 11 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:320]: We need file_size so that we know how many characters of file data we are supposed to receive from the crashed device.
We get this file size from the device itself after we pass the name of the file that we want.
The device side performs a stat for us to get the file size.
However, the device sends back one character at a time using up_lowputc, but we need 8 characters (64 bytes) in total for file_size.
Hence, we receive one character at a time and combine them using this logic into a 64 bit file_size.

We cannot perform a stat to get the userfs file size here (in trap.c) because this file is run on the Linux host machine

File: tools/trap/trap.c (hunk 13 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:533]: 1. I don't think we should ask the user for the size of name length if we assume that the user is not familiar with our system. They might not know about the concept of a MAX NAME LENGTH in this case.
2. I agree with this point, we should use a common config, something maybe like CONFIG_FS_MAXNAMLEN. The maximum file path length should be a common restriction on any file system that passes through vfs.
As of now, we only have the one file system i.e. jSmartFs
3. I get your point. We can use an arbitrary limit on file name length here, and when the file name is passed to the device for dumping, the device checks the length of the file name and returns an error if it is too long (ENAMETOOLONG)
Let me check this third possibility

File: os/board/common/crashdump.c [hunk 16]
```diff
@@ -208,6 +210,68 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
 
+static int send_file(void)
+{
+	int i;
+	int fd;
+	int ret;
+	struct stat st;
+	off_t size;
+	char buf;
+	char filename[CONFIG_SMARTFS_MAXNAMLEN + 1];
```
Comment (kishore-sn) [inline os/board/common/crashdump.c:221]: I think its better to initialize the filename with null characters.

File: os/board/common/crashdump.c [hunk 17]
```diff
@@ -208,6 +210,68 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+
+	ret = stat(filename, &st);
+	if (ret != OK) {
+		/* We are not able to stat the file for some reason */
+		return ERROR;
+	}
+	size = st.st_size;
+
+	/* This Ack signal confirms the presence of the requested file */
+	//up_lowputc('A');
```
Comment (kishore-sn) [inline os/board/common/crashdump.c:243]: If not required, lets delete these lines.

File: os/board/common/crashdump.c (hunk 10 above)
Comment (kishore-sn) [reply os/board/common/crashdump.c:260]: In the crashed state, we dont know what is the state of kernel. So, all kernel operations must be avoided as much as possible. So, let us avoid malloc. However, if reading one character at a time from the file causes too much overhead, then we could maintain a static buffer of sectorsize or blocksize. It is left to you to experiment with this and decide if it is required.

File: tools/trap/trap.c [hunk 18]
```diff
@@ -288,6 +294,67 @@ static int extfsdump_recv(int dev_fd)
...
+	file_size = 0;
+	for (int i = 0; i < 8; i++) {
+		ret = read(dev_fd, &char_filesize[i], sizeof(char_filesize[i]));
+		if (ret != sizeof(char_filesize[i])) {
+			printf("Unable to get file size from Target Device, ret = %d\n", ret);
+			return -1;
+		}
+	}
+
+	file_size = (((u_int64_t)char_filesize[0] << 56) & 0xFF00000000000000U) | (((u_int64_t)char_filesize[1] << 48) & 0x00FF000000000000U) | (((u_int64_t)char_filesize[2] << 40) & 0x0000FF0000000000U) | (((u_int64_t)char_filesize[3] << 32) & 0x000000FF00000000U) | ((char_filesize[4] << 24) & 0x00000000FF000000U) | ((char_filesize[5] << 16) & 0x0000000000FF0000U) | ((char_filesize[6] <<  8) & 0x000000000000FF00U) | (char_filesize[7] & 0x00000000000000FFU);
```
Comment (kishore-sn) [inline tools/trap/trap.c:325]: This must work. Instead of having such long statement.
```
for (int i = 0; i < 8; i++)
    filesize |= (u_int64_t)char_filesize[8 - i - 1] << (i * 8);
```

File: os/board/common/crashdump.c (hunk 16 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c:221]: Understood, will do

File: tools/trap/trap.c (hunk 18 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:325]: Yes will consider this, thankyou

File: tools/trap/trap.c [hunk 19]
```diff
@@ -288,6 +296,70 @@ static int extfsdump_recv(int dev_fd)
...
+	for (i = 0; i < 8; i++) {
+		ret = read(dev_fd, &char_filesize[i], sizeof(char_filesize[i]));
+		if (ret != sizeof(char_filesize[i])) {
+			printf("Unable to get file size from Target Device, ret = %d\n", ret);
+			return -1;
+		}
+	}
+
+//	file_size = (((u_int64_t)char_filesize[0] << 56) & 0xFF00000000000000U) | (((u_int64_t)char_filesize[1] << 48) & 0x00FF000000000000U) | (((u_int64_t)char_filesize[2] << 40) & 0x0000FF0000000000U) | (((u_int64_t)char_filesize[3] << 32) & 0x000000FF00000000U) | ((char_filesize[4] << 24) & 0x00000000FF000000U) | ((char_filesize[5] << 16) & 0x0000000000FF0000U) | ((char_filesize[6] <<  8) & 0x000000000000FF00U) | (char_filesize[7] & 0x00000000000000FFU);
+	for (i = 7; i >= 0; i--) {
```
Comment (amogh-samsung) [inline tools/trap/trap.c:328]: Pending resolution

File: tools/trap/trap.c [hunk 20]
```diff
@@ -367,7 +367,7 @@ int main(int argc, char *argv[])
 	char *dev_file;
 	char tty_path[TTYPATH_LEN] = {0, };
 	char tty_type[TTYTYPE_LEN] = {0, };
-	int choice;
+	char choice;
```
Comment (Tirtha-Samsung) [inline tools/trap/trap.c:370]: Why did you change the int to char? is it needed?

Comment (amogh-samsung) [reply tools/trap/trap.c:370]: Minor change to facilitate accepting 'r' and 'x' as inputs for resetting and exiting.

File: os/board/common/crashdump.c [hunk 21]
```diff
@@ -208,6 +210,68 @@ static int get_ext_userfs_part(size_t *size, int *index)
...
+	char filename[CONFIG_SMARTFS_MAXNAMLEN + 1];
+	char char_filesize;
+
+	/* Send acknowledgement for USERFS FILE DUMP Handshake */
+	up_lowputc('A');
+
+	/* Receive the file name to be sent to the host */
+	for (i = 0; i < CONFIG_SMARTFS_MAXNAMLEN + 1; i++) {
+		filename[i] = '\0';
+	}
```
Comment (sunghan-chang) [inline os/board/common/crashdump.c:230]: Is this necessary?
If necessary, how about doing at definition?
```
char filename[CONFIG_... + 1] = { 0, };
```

File: tools/trap/trap.c [hunk 22]
```diff
@@ -24,6 +24,14 @@
...
 
 #include "trap.h"
+#include "../../os/include/tinyara/config.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#ifndef CONFIG_SMARTFS_MAXNAMLEN
+#define USERFS_MAXNAMLEN 32
+#endif
```
Comment (sunghan-chang) [inline tools/trap/trap.c:34]: Even you include the header, is this necessary?
If you think some config which does not have the config in the .config, it could be fs disabled.

File: tools/trap/trap.c [hunk 23]
```diff
@@ -350,32 +421,35 @@ int main(int argc, char *argv[])
...
 			DUMP_FLAGS |= RAMDUMP_FLAG;
 			DUMP_FLAGS |= USERFSDUMP_FLAG;
 			break;
-		case 4:
+		case '4':
 			DUMP_FLAGS |= EXTUSERFS_DUMP_FLAG;
 			break;
-		case 5:
+		case '5':
+			DUMP_FLAGS |= FILE_DUMP_FLAG;
```
Comment (sunghan-chang) [inline tools/trap/trap.c:446]: Does this support multiple command at once? If not, we can just set it without bit comparision.

File: os/board/common/crashdump.c (hunk 21 above)
Comment (amogh-samsung) [reply os/board/common/crashdump.c:230]: Agreed

File: tools/trap/trap.c (hunk 22 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:34]: In case fs is not disabled
The application will return an error while looking for the file that the user has asked to dump.

Also, I believe that the user may be using userfs with SmartFS enabled but may not build the code themselves and may not have the .config file available.
Hence, there are other ways to provide the parameter of maximum file name length.

File: tools/trap/trap.c (hunk 23 above)
Comment (amogh-samsung) [reply tools/trap/trap.c:446]: Does not support currently,
We can do it
I will add this in a separate commit.

**Comment by amogh-samsung** [comment]
> Sorry for late comment. Is it necessary at assert case to get file dump? File data is in the flash so that it will not disappear after reset.

We do not prefer having an assert case for dumping FS contents. Ideally, we should be able to dump them without crashing the device.
However, when the device is active, we might need a separate UART-console communication channel to transfer the dump along with the usual console logs.
@kishore-sn could you help me explain this point please?

Additionally, @sunghan-chang , I have wanted to make FS dump available without crashing and I will push commits to make it work if and when I do find a way.

**Comment by sunghan-chang** [comment]
@amogh-samsung I think we can use a new TASH command for that. Impossible?

**Comment by amogh-samsung** [comment]
> @amogh-samsung I think we can use a new TASH command for that. Impossible?

I would like to check that possibility, but would also like to have this feature included here in the TRAP tool.
What do you suggest?

**Comment by sunghan-chang** [comment]
It becomes much huge file (over 500 lines!). Big file is much difficult to analyse and to modify later so that I want to confirm whether this is really necessary or not. I see no one use this tool even though we added several dump feature..

**Comment by amogh-samsung** [comment]
> It becomes much huge file (over 500 lines!). Big file is much difficult to analyse and to modify later so that I want to confirm whether this is really necessary or not. I see no one use this tool even though we added several dump feature..

The intent of this change is a mixture of 'good-to-have' and 'debugging-utility'
We have used this feature internally in the past to test and debug certain other implementations.
@r-prabu could probably help explain further

File: os/board/common/crashdump.c [hunk 24]
```diff
@@ -208,6 +210,65 @@ static int get_ext_userfs_part(size_t *size, int *index)
 	return 0;
 }
 
+static int send_file(void)
+{
+	int i;
+	int fd;
+	int ret;
+	struct stat st;
```
Comment (pillip8282) [inline os/board/common/crashdump.c:218]: How about to initialize values to remove initialization warning in code review tool.

**Comment by r-prabu** [comment]
> tool even though we added several dump feature

@sunghan-chang , I think Users are not aware of the options available. even in SRIB, a team from DA (WAC) mentioned this is useful during development. Let us know your opinion on this. Also this PR is kept open for a long time. If you want to change the command order we can take it in separate PR immediately after merging this.

**Review by sunghan-chang — APPROVED** [review]

**Review by r-prabu — APPROVED** [review]
